File: backend/app/schemas/analysis.py
from typing import Any, Literal

from pydantic import BaseModel, Field

from app.schemas.deployment_recommendation import (
    DeploymentRecommendationResponse,
)


class HardConstraintResult(BaseModel):
    passed: bool
    status: str
    reason: str


class HardConstraintsResponse(BaseModel):
    passed: bool
    constraints: dict[str, HardConstraintResult]
    failed_constraints: list[str]


class SoftConstraintResult(BaseModel):
    score: float | None
    value: float | None
    unit: str


class SoftConstraintsResponse(BaseModel):
    score: float
    constraints: dict[str, SoftConstraintResult]


class TechnicalFeasibilityResponse(BaseModel):
    is_feasible: bool
    feasibility_score: float
    decision: str
    hard_constraints: HardConstraintsResponse
    soft_constraints: SoftConstraintsResponse
    constraint_summary: str


class AnalysisRequest(BaseModel):
    latitude: float = Field(..., ge=-90, le=90)
    longitude: float = Field(..., ge=-180, le=180)

    land_area_hectares: float = Field(..., gt=0)
    available_budget: float = Field(..., gt=0)

    location_name: str | None = Field(
        default=None,
        max_length=255,
        description="Human-readable site location name.",
    )

    # Optional for backward compatibility with existing API clients/tests.
    # New Analysis UI always sends these values.
    project_type: Literal["solar", "wind", "hybrid"] = Field(
        default="hybrid",
        description="Requested project type: solar, wind, or hybrid",
    )
    installation_type: Literal[
        "ground-mounted",
        "rooftop",
        "other",
    ] = Field(
        default="ground-mounted",
        description="Installation configuration: ground-mounted, rooftop, or other",
    )


class AnalysisResponse(BaseModel):
    solar_features: dict
    wind_assessment: dict

    renewable_score: float
    terrain_score: float
    infrastructure_score: float
    environmental_score: float
    economic_score: float

    overall_site_score: float
    ml_prediction: dict

    technical_feasibility: TechnicalFeasibilityResponse

    # Standardized Day-29 response fields
    site_suitability: float
    recommended_deployment: str
    energy_yield: dict
    financial_metrics: dict
    recommendation_reason: str

    # Dynamic deployment recommendation.
    # Exposes priority, confidence, deployment type and reason
    # directly to frontend consumers.
    recommendation: DeploymentRecommendationResponse

    # Extended intelligence
    deployment_plan: dict
    sentinel2: dict | None = None

File: frontend/src/lib/api/analysis.ts
import { apiRequest } from "./client";

export type AnalysisRequest = {
  latitude: number;
  longitude: number;
  land_area_hectares: number;
  available_budget: number;
  location_name?: string | null;
  project_type?: "solar" | "wind" | "hybrid";
  installation_type?: "ground-mounted" | "rooftop" | "other";
};

export type ConstraintResult = {
  passed?: boolean;
  status?: string;
  reason?: string;
  score?: number;
  value?: number;
  unit?: string;
  slope?: number;
  [key: string]: unknown;
};

export type FeasibilityConstraints = {
  terrain?: ConstraintResult;
  land_use?: ConstraintResult;
  grid_proximity?: ConstraintResult;
  road_accessibility?: ConstraintResult;
  [key: string]: ConstraintResult | undefined;
};

export type TechnicalFeasibility = {
  is_feasible: boolean;
  feasibility_score: number;
  decision: string;
  hard_constraints?: {
    passed?: boolean;
    constraints?: FeasibilityConstraints;
    failed_constraints?: string[];
  };
  soft_constraints?: {
    score?: number;
    constraints?: FeasibilityConstraints;
  };
  constraint_summary?: string;
};

export type EnergyYield = {
  solar_energy?: number;
  wind_energy?: number;
  total_energy?: number;
  deployment_type?: string;
};

export type FinancialMetrics = {
  annual_revenue?: number;
  estimated_project_cost?: number;
  payback_period?: number;
  roi?: number;
};

export type Recommendation = {
  deployment_type?: string;
  confidence?: number;
  priority?: string;
  reason?: string;
  generated_at?: string;
};

export type Optimization = {
  recommended_technology?: string;
  recommended_capacity_mw?: number;
  expansion_status?: string;
  [key: string]: unknown;
};

export type DeploymentPlan = {
  recommendation?: Recommendation;
  recommended_capacity_mw?: number;
  expansion_status?: string;
  energy_estimation?: EnergyYield;
  financial_analysis?: FinancialMetrics;
  optimization?: Optimization;
  [key: string]: unknown;
};

export type MlFeature = {
  feature: string;
  importance: number;
};

export type MlPrediction = {
  solar_pvout_potential?: number;
  explanation?: {
    top_features?: MlFeature[];
    summary?: string;
  };
  [key: string]: unknown;
};

export type AnalysisResponse = {
  solar_features: {
    solar_irradiance: number;
    temperature: number;
    relative_humidity: number;
    [key: string]: unknown;
  };

  wind_assessment: {
    wind_speed: number;
    classification: string;
    capacity_factor: number;
    [key: string]: unknown;
  };

  renewable_score: number;
  terrain_score: number;
  infrastructure_score: number;
  environmental_score: number;
  economic_score: number;
  overall_site_score: number;

  ml_prediction: MlPrediction;
  technical_feasibility: TechnicalFeasibility;

  site_suitability: number;
  recommended_deployment: string;

  energy_yield: EnergyYield;
  financial_metrics: FinancialMetrics;

  recommendation_reason: string;

  deployment_plan: DeploymentPlan;

  sentinel2?: Record<string, unknown> | null;
};

export function runAnalysis(
  request: AnalysisRequest,
): Promise<AnalysisResponse> {
  return apiRequest<AnalysisResponse>("/analysis", {
    method: "POST",
    body: JSON.stringify(request),
  });
}
